### DISCUSSION TITLE
State of React Native 2023 Preview / Call for Questions

### DISCUSSION THREAD

**Discussion by bbarthec** [software-mansion/state-of-react-native#6] [Preview / Call for Questions]
## [Live preview of the survey](https://preview.stateofreactnative.com/)

At Software Mansion, we've iterated on last year's edition and improved the question set and we're still improving it at this point! Nevertheless, we invite you to check out the preview version. If there's still something missing and/or unnecessarily added, let us know. Don't forget to justify your suggestions!

We'll do our best to review the suggestions on a daily basis and respond accordingly by either accepting or rejecting them.

The preview period will last until the last week of November. The survey will go public on the 4th of December (first Monday in December) and would run until the 5th of January 2024 (first Friday in January). 

The results will be published after the survey has closed. Expected date: second half of January 2024.

## Meaningful changes we've introduced this year

```diff
Developer background
  Target platforms
+   React Native Skia (Linux & macOS)

  What mobile OS versions do you target
-   Most recent ones (iOS 16+, Android 13+)
-   Reasonably current (iOS 12+, Android 10+)
-   Old (iOS 10+, Android 7+)
+   Most recent ones (iOS 17+, Android 14+)
+   Reasonably current (iOS 15+, Android 12+)
+   Old (iOS 12+, Android 9+) 

Platforms APIs
  Platform APIs usage
+   Foldable
+   Live Activity (Dynamic Island)

State management
- Legend State     # too niche
- React Easy State # too niche
+ MobX             
+ MobX-State-Tree  
+ XState           
+ Redux Toolkit

Data fetching
+ RTK Query

Styling
+ Dripsy   # moved here from UI Components libraries
+ react-native-unistyles
+ Emotion
+ React Native Zephyr
+ gluestack-style
+ Fast Styles
+ Tailwind React Native Classnames (twrnc)

UI Components libraries
- Dripsy # moved to Styling
+ gluestack-ui
+ Tamagui
+ react-native-material-kit

Graphics and animations
- Legend Motion  # too niche
+ Rive

Other app aspects
  Storage
-   RxDB    # too niche
  
  Data Visualisation 
-   React Native Charts Wrapper   # too niche

- Interactions          # too few options, combined into new "Other popular libraries"

- Internationalization  # too few options, combined into new "Other popular libraries"
  
  Analytics
- Flurry   # too niche

  Crash reporting
- App Dynamics          # too niche
- Amazon CloudWatch     # too niche
- Elastic Observability # too niche

  Testing
- XCUITest  # too niche
- Espresso  # too niche
+ Storybook

+ Other popular libraries
+  @expo/vector-icons
+  @gorhom/bottom-sheet
+  @shopify/flash-list
+  expo-image
+  react-i18next
+  react-native-camera
+  react-native-fast-image
+  react-native-gesture-handler
+  react-native-localize
+  react-native-maps
+  react-native-safe-area-context
+  react-native-vision-camera
+  react-native-webview

+ Other popular libraries not listed above

React Native features
  React Native version
+   nightly
+   next
+   0.73
+   0.72

+ Bridgeless Mode

+ React Native Directory # moved here from "Learning methods"

React Native tools
  Started templates
+   create-react-native-app

+ Upgrade Helper

+ Config Plugins

- Monorepo tools   # this is nicely covered by "State of JS" survey

React Native Alternatives
- Felgo  # too niche

Resources
  Blogs & magazines
# most popular responses from last year edition
+   Callstack blog
+   DEV Community (dev.to)
+   Expo blog (Exposition)
+   Kent C. Dodds
+   LogRocket
+   Medium
+   Microsoft React Native blog
+   r/reactnative
+   React Native blog
+   React Native Newsletter (by Infinite Red)
+   Software Mansion blog
+   This Week In React (by Sébastien Lorber)

  Sites & courses
# most popular responses from last year edition
+   Animate React Native (by Catalin Miron)
+   Codecademy
+   Developer training (by Callstack)
+   Egghead.io
+   Epic React (by Kent C. Dodds)
+   Expo docs
+   Frontend Masters
+   Medium
+   React Native docs
+   Reddit
+   Stack Overflow
+   Start React Native (by William Candillon)
+   Udemy
+   X (Twitter)
+   YouTube

  Podcasts
# most popular responses from last year edition
+   PodRocker (by LogRocket)
+   React Native Radio (by Infinite Red)
+   React Native Show (by Callstack)
+   Syntax.fm
+   The Changelog

- Online Communities  # too niche

Opinions
  React Native pain points
# most popular responses from last year edition
+   App Deployment
+   App Building
+   Dealing with native code
+   Debugging
+   Documentation
+   Unusable error messages
+   Graphics and animations
+   Keyboard handling
+   Navigation solutions
+   New architecture
+   Performance
+   Platforms inconsistencies
+   Project setup
+   Reliance of 3rd party modules
+   Stability
+   Styling limitations
+   Unmaintained packages
+   Upgrades

  Missing features
# most popular responses from last year edition
+   Android shadows
+   Better animations/gestures
+   Better debugging
+   Better profiling
+   Better documentation
+   Better FlatList
+   Better keyboard handling
+   Better performance
+   Better testing
+   Better Web compatibility
+   CSS features
+   First-party libraries
+   Multithreading
+   SwiftUI/Jetpack Compose integration

React Native advantages
# most popular responses from last year edition
+   Code sharing
+   Community
+   Cost-effective
+   Cross-platform
+   Developer experience
+   Documentation
+   Easy to learn
+   Ecosystem
+   Expo
+   Fast iterations
+   Hot reloading
+   Native feel of the app
+   Over-the-air updates
+   Performance
+   Single codebase
+   Stable
+   Time to market
+   Web technologies
```

**Comment by tido64** [comment]
Can we add [react-native-test-app](https://github.com/microsoft/react-native-test-app) under "React Native tools"?

**Reply by bbarthec** [reply]
Yeah, I see it as another question with 👍/👎/✅/🚫/🤷 answers 👀

**Comment by matthargett** [comment]
Apollo Client under state tools? Apollo Local State was used in >50% of RNPS apps, and we were actively migrating to it last year in our RN port to Lua at Roblox.

**Reply by bbarthec** [reply]
I'm not sure I'm following your suggestion here 🤔 
Apollo Client is listed as the first option in both the `State management`  and `Data fetching` sections. I read in the docs that [`Apollo Local State` is a part of `Apollo Client`](https://www.apollographql.com/docs/react/local-state/local-state-management). Is there anything else that I'm missing here? 🤔 
| State management | Data fetching |
|-|-|
| ![Screenshot 2023-11-07 at 10 02 07](https://github.com/software-mansion/state-of-react-native/assets/16623003/5b7593f2-808a-4e37-a3f2-f0d16ff0cdc8) | ![Screenshot 2023-11-07 at 10 04 42](https://github.com/software-mansion/state-of-react-native/assets/16623003/a7a1f0de-4e99-49e4-a3ee-bc62e690fdd4) |

**Comment by vargajacint** [comment]
Can we add [Rocket Ship](https://podcasts.apple.com/us/podcast/rocket-ship/id1700257644) to podcasts?

**Reply by bbarthec** [reply]
Yeah, sure 👍 We'll add that one 🙂

**Comment by thymikee** [comment]
For Resources, does it make sense to include specific books, like ["The Ultimate Guide to React Native Optimization"](https://www.callstack.com/ebook/the-ultimate-guide-to-react-native-optimization)?

React Native Tools |> Package managers: I'd add Bun
React Native Tools: Since there is Expo CLI and Config Plugins, how about Expo Go and Custom Development Client?

**Reply by bbarthec** [reply]
["The Ultimate Guide to React Native Optimization"](https://www.callstack.com/ebook/the-ultimate-guide-to-react-native-optimization) - we'll convert `Learning Methods > Sites & courses` into `Sites, courses & books` and we'll list it there 👍 

`bun` - we'll add that one 👍 

`Expo Go` & `Custom Development Client` - yeah, these sound like some obvious tools we should ask about. We'll add them 😊

**Comment by grahammendick** [comment]
Thanks for including my navigation library in the survey. Could you call it "The Navigation router" instead of "navigation-react-native", please? (navigation-react-native is just the name of one of the npm packages that make up the Navigation router).

**Reply by bbarthec** [reply]
Yeah, sure thing! We'll change that 😊

**Comment by jbroma** [comment]
Can we add [Re.Pack](https://github.com/callstack/repack) under "React-Native tools"?

**Reply by bbarthec** [reply]
At the moment we don't really ask about bundlers or bundler-related solutions anywhere in the survey. If we were to do so, we would need to create a new section on this specific topic and consider adding other solutions in this category. And for the sake of brevity and the fact that the survey is already quite long, we would rather not add another section this year.

Alternatively what do you say about creating a single question about bundlers and bundler-related solutions in the `React Native features` section? The same way we ask about `OTA updates solutions`. 

Do you mind suggesting other answers for such a question?

**Comment by keeslinp** [comment]
How about platform colors under styling?

**Reply by bbarthec** [reply]
You suggest adding [PlatformColor API](https://reactnative.dev/docs/platformcolor) as an option in this section, right? 🤔 This section is more about the solutions used for styling the app rather than the specific React Native APIs. If we were to add such an option in this section, we would also have to consider adding other React Native APIs in other sections, or creating a whole new one, and we have refrained from doing so.
TL;DR; we would rather not add that one

**Reply by keeslinp** [reply]
That makes sense. I'm mostly curious cause most of the common styling/ui libraries in react native have specifically said they don't want to support it so I would just want to know where the ecosystem stands. Platform colors solves a core feature that most of those libraries are implementing themselves (dark mode)

**Comment by nandorojo** [comment]
Can we please add:

1. [Solito](https://solito.dev) under **Navigation**  
2. "Solito Docs" under **Resources → Sites & courses**
3. `create-solito-app` under **React Native Tools → Starter templates**
4. [Burnt](https://github.com/nandorojo/burnt) and [Zeego](https://zeego.dev) under **Other app aspects → other popular libraries**

**Reply by bbarthec** [reply]
1. `Solito` under navigation - sure thing! We were considering that already! 😁 
2. `Solito docs` - okay, that sounds reasonable 👍 
3. `create-solito-app` - yeah, we'll enrich this section with more options including this one 😊 
4. `Burnt` and `Zeego` - too niche in our opinion to be included in this section 😞 We considered 50k weekly downloads as our popularity threshold, otherwise this section would grow massively 😱

**Comment by flexbox** [comment]
I think we could have a dedicated Starter template section

- `react-native-cli`
- `create-expo-app`
- `ignite-cli`
- `create-solito-app`
- ...

and some other community resources listed here https://weshipit.today/react-native-starters

**Reply by bbarthec** [reply]
Yeah, we'll enrich this question with these suggestions! 👍

**Reply by bbarthec** [reply]
Oh, I've misunderstood your suggestion first! Having a separate section fot this topic is too much right 😞 We've refrained from adding another section to the survey this year.

**Reply by flexbox** [reply]
No problem, I am a big fan of less is more.
maybe next year 😉

**Comment by ignaciosantise** [comment]
What do you think of adding [Zustand](https://github.com/pmndrs/zustand) and [Valtio](https://github.com/pmndrs/valtio) under **State management** ?

**Reply by bbarthec** [reply]
`Zustand` is already there 😉
Last year `Valtio` was only mentioned 4 times in the [`Other state management` question](https://results.stateofreactnative.com/state-management). If it's mentioned a lot this year, we'll add it to the pre-defined answers next year 😉

**Comment by SachaG** [comment]
Just wanted to drop by to say we've improved the Devographics codebase a *lot* since last year, it might be worth taking a look if you can merge in the changes!

**Reply by kacperkapusciak** [reply]
Hi @SachaG  👋

Thank for the heads up! We're gathering ourselves to do exactly that but considering the amount of hacks and workaround we've added to the current codebase there's a possibility we may not make it in time for State of React Native 2023

Keep fingers crossed! 🤞
And, thank you for your work on the surveys 🙇

**Comment by grahammendick** [comment]
You shouldn’t include the react router in the Navigation section. Move it into “Other popular libraries” like you did with libraries that have too few options to justify their place in the dedicated section. The react router doesn’t even have a stack where visited scenes are held in memory. A stack should be the lowest bar for entry into the Navigation section.

**Reply by kacperkapusciak** [reply]
We'll going to keep it in this year's survey as it had considerable amount of responses last year

Altho it's a sad navigation without any animations, tabs, drawers some developers still use it. Probably by some legacy codebase constraints I guess 🤷

**Reply by grahammendick** [reply]
I think developers use it because it's blessed by the community. For example, it was recommended in the Expo documentation for a long time. Removing it from this survey would be an important step in educating developers that it isn't a good choice. If you need more convincing, the [react router native package is about 30 lines of code](https://github.com/remix-run/react-router/blob/main/packages/react-router-native/index.tsx)

**Comment by louiszawadzki** [comment]
That might be too niche, but I'd be interested to know in "Testing" how many people write native tests (in kotlin/java and swift/obj-c), what do you think?

Also I didn't see categories for:
- Debugging (console.log, Flipper, Chrome devTools, XCode, Android Studio, etc.)
- Performance monitoring (probably same choices as for Crash Reporting, so maybe including it in the name as "Crash Reporting and Performance Monitoring"?)

Did you consider these categories as well?

In any case, thanks for the awesome work of putting all this together year after year, this is incredible work!!

**Reply by bbarthec** [reply]
@louiszawadzki, right now we cover these aspects in `React Native features` section, but we'll be reiterating over it again before going public 💪 
![Screenshot 2023-11-22 at 10 38 57](https://github.com/software-mansion/state-of-react-native/assets/16623003/3fec7bc6-d868-4cef-b1ef-20839d7c4d85)
![Screenshot 2023-11-22 at 10 38 54](https://github.com/software-mansion/state-of-react-native/assets/16623003/f07a43ae-d544-4aab-ab81-68288b845508)

**Reply by bbarthec** [reply]
> That might be too niche, but I'd be interested to know in "Testing" how many people write native tests (in kotlin/java and swift/obj-c), what do you think?

@louiszawadzki, I've realised I haven't replied to this part of your message 🤦 Sorry for that 🙇  
I think it's too niche indeed 🤔 If someone is doing some native testing I expect that answering the `Other testing solutions` question might be the best option 😉

**Reply by louiszawadzki** [reply]
Thanks for taking the time to answer all the questions 🙇 
And thanks for pointing the correct sections 👍

**Comment by Okelm** [comment]
@bbarthec how about [react-native-video](https://www.npmjs.com/package/react-native-video) under "Other app aspects → other popular libraries"? Much more than 50k weekly downloads threshold :)

**Reply by bbarthec** [reply]
@Okelm, yeah, we'll add that 👍

**Comment by huntie** [comment]
@bbarthec It would be fantastic to get more signal on Debugging. Here's an expanded question set that I've prepared with the team. Happy to discuss details here or over Discord.

Source drawing: [link (download and open with Excalidraw)](https://gist.githubusercontent.com/huntie/311eb18d5f41e97eacf3e31b4d7e05f0/raw/1e5efc3c1312c02a87592654fb94afbf839ec4d0/RN%2520Survey%2520Debugging%2520additions.excalidraw).

<img alt="RN Survey Debugging additions" width="200" src="https://github.com/software-mansion/state-of-react-native/assets/2547783/356d2411-b781-4ba9-a698-4e764b14eaf9">

**Reply by bbarthec** [reply]
@huntie, we've discussed your suggestion and have decided to add another section to the survey called `Debugging and profiling`. We'll include all the suggestions there except the second one  `Primary debugging tools` 💪 (In other sections, we don't ask about the `primary` choice 🤷)

**Reply by huntie** [reply]
Sounds perfect. Thanks for taking these suggestions and for actioning so fast!

**Reply by byCedric** [reply]
I did a pass on this too, thanks for adding this @huntie & @bbarthec 😄 Here are my 2c:

### Debugging tools
Not sure if I understand the "Expo/Hermes" and "New Debugger" separation. Both are powered by the Hermes debugger part, it's just a slightly different UI.

IMHO, since the "New Debugger" still has to be released (0.73 -- non-rc), and SDK 50 likely will use this new debugger, it doesn't make a lot of sense to me to split this one up. I think just having "Chrome DevTools", with a possible other question asking the user what they are using (Expo vs RNC) might give the same results without confusing users too much.

### JavaScript debugging features (Chrome DevTools)
The "web only" and "expo only" seems a bit unnecessary here. The question asks the user about "Which Chrome DevTools features do you use (across current RN debugging tools and/or for web development)". 

IMHO, it might actually affect how users answer this question, we can do without. We know what is implemented/available. (Plus, React DevTools is available in Expo, and Network inspector is ofc implemented on web)

### Performance profiling
Might be nice to add [Flashlight](https://github.com/bamlab/flashlight) here too. Announced at App.js and continuously improved by BAM lab since. Seems to be something users should know about, and might be nice to gather stats about.

### Debugger clients?
Does it make sense to add another question that compares the debugger "clients" used? E.g. are people using VSCode ([vscode-react-native](https://github.com/microsoft/vscode-react-native#readme) / [vscode-expo-tools](https://github.com/expo/vscode-expo#readme)) or just chrome devtools?

Hope it helps!

**Reply by huntie** [reply]
Thanks Cedric, let's take on all these suggestions 💯. My notes:

> Debugging tools

Yeah, 100% agree the end state will be less fragmented. One thing is we didn't keep "Chrome DevTools" in our branding for the New Debugger, which could make that unclear. But with the finalised timing of this survey, it makes sense to remove it entirely.

Let's have:

- Chrome DevTools (in browser — Remote JS Debugging)
- Chrome DevTools (Hermes)
- Chrome DevTools (Expo)
- [others, beginning with "Console logs"]

(Hermes and Expo are overlapping, but I reckon this makes these options less ambiguous for the avg developer — and the Expo debugger has meaningfully expanded features today.)

> JavaScript debugging features (Chrome DevTools)

Yeah I guess it was with the aim to clarify up front that a number of these features are only available in some places.

**Reply by bbarthec** [reply]
@byCedric, thank you a ton for taking a look 👏 

> ### Debugging tools
> (...)

Yeah, I agree with both you and @huntie, so we will have following list for `Debugging tools` question:
- `Chrome DevTools (in browser — Remote JS Debugging)`
- `Chrome DevTools (Hermes)`
- `Chrome DevTools (Expo)`
- `Console logs`
- `Flipper`
- `Reactotron`
- `React DevTools`
- `React Native Debugger`
- `Native debugging (Xcode / Android Studio)`

We also need to refactor `Problems - debugging tools`:
- `Connecting debugger to application - Expo/Hermes debugger` ➡️ we should split it into individual answer for `Expo` and `Hermes`
- should we add `Connecting debugger to application - Chrome DevTools`?
- should we also add similar ☝️ answers for other non-core-ReactNative-, non-Meta- or non-Expo-related technologies, like `Connecting debugger to application - Reactotron`/` - React Native Debugger`? 
- should we shorten the phrase `Connecting debugger to application - (..)` into `Connection debugger - (...)` or even replace it with `Using debugger - (...)`? cc @huntie

> ### JavaScript debugging features (Chrome DevTools)
> (...)

I agree, let's drop specifics like `web only`/`Expo only`

> ### Performance profiling
> (...)

Yeah, `Flashlight` is to be added 💯 

> ### Debugger clients
> (...)

How do you see this question? In the `Debugging and profiling` section? As a multiselect with checkboxes or as a 👍/👎/✅/🚫/🤷 question per technology?
As far as I can tell we can identify two `vscode` plugins that aren't mentioned in the survey so far and nothing more? 🤔 
Given that, I think we want to preserve the idea of keeping `vscode`-related technologies away from this survey this year 🤔

**Reply by huntie** [reply]
> We also need to refactor `Problems - debugging tools`:

Let's shorten to "Connecting debugger to app — [type]", where type is aligned with the previous changes (per "Debugging tools") + "React DevTools". Up to you on inclusion of the others.

- Connecting debugger to app - Flipper
- Connecting debugger to app - Chrome DevTools (Hermes)
- Connecting debugger to app - Chrome DevTools (Expo)
- Connecting debugger to app - React DevTools
- ...

"Chrome DevTools (in browser — Remote JS Debugging)" is intentionally omitted.

**Comment by bbarthec** [comment]
### Heads up folks ⚠️ 

### The preview ends on Thursday, 30th of November

So far, all the suggested changes reported in this discussion or via the Discord have been addressed and the vast majority of them landed in the survey 🎉 

This is the last call for any changes you would like to see in the survey.

### Reminder: **The survey goes public on Monday 4th of December 🚀**

**Comment by elicwhite** [comment]
I don't see something that would give us signal about code push usage vs expo-updates. Is it in there somewhere I missed?

On the react-native version question, my sense is "which version do you currently use" is helpful in connection with the other questions in the survey, but what I'm really interested in is "what is your strategy for upgrading to new versions". With something like: Stay on latest, don't upgrade until I have to, upgrade occasionally, upgrade for new features, freeform

I also think the About You section of the survey is super important to understand the demographics of the audience answering the questions, and I'm a little nervous that due to the length of the survey most respondents won't get that far, or the demographics of the people who /do/ get that far will by itself significantly skew the results. Is it possible to move it earlier? WDYT?

**Reply by kelset** [reply]
(about the last point, IIRC some surveys put it at the very start precisely to avoid people not answering)

**Reply by bbarthec** [reply]
@TheSavior, thank you for sharing your point of view 👏 

> I don't see something that would give us signal about code push usage vs expo-updates. Is it in there somewhere I missed?

We're asking about it in `React Native features` section 💪 
![Screenshot 2023-11-29 at 16 04 36](https://github.com/software-mansion/state-of-react-native/assets/16623003/ff2ca90f-18a3-4d81-a2bc-e551f990ca4f)

> `(...) "what is your strategy for upgrading to new versions" (...)`

I like that one 👏 
So the proposed answers could be:
- `stay on the latest`
- `upgrade occasionally`
- `upgrade for new features`
- `upgrade in parallel with Expo SDK releases`, cc @TheSavior, @brentvatne, @byCedric, @tsapeta what do you think about such option?
- `don't upgrade until I have to`
- `freeform`

> I also think the About You section of the survey is super important to understand the demographics of the audience answering the questions, and I'm a little nervous that due to the length of the survey most respondents won't get that far, or the demographics of the people who /do/ get that far will by itself significantly skew the results. Is it possible to move it earlier? WDYT?

I understand the importance of this part very well 👍 
At the moment, we're copying the same approach as Devographics's surveys ([State of JS](https://stateofjs.com), [State of CSS](https://www.stateofcss.com), [State of React](https://stateofreact.com), etc.). They've been ordering the sections this way for a couple of editions now and, although I can't find a reference right now, I think it's because the survey is more about the technology and less about the people, so we endorse technology-related questions first 🤔
Nevertheless, last year edition completeness per section data is ([State of React Native 2022 results](https://results.stateofreactnative.com/)):
- `About you` - ~65% - it was also the last section in the pervious edition of the survey, but was placed first on the results page
- `Developer background` - ~98%
  - first section asked last year
  - first major part of respondents drops after completing this section
  - the completion ratio for this section is rather skewed anyway
    - we had 1366 anonymous respondents and 521 logged-in respondents
    - these anonymous respondents partially consist of people that started the survey more than once (e.g. when I was filling the State of JS survey, I interrupted the process after the first section and when I tied to resume it later I was starting from the scratch - this might happen when the responder is resuming (starting) the survey on different device or browser or when the responder decide to became logged-in responder (my case))
- `Platform APIs`- 84% - organic churn starts from this section onwards
- `State management` - ~84%
- `Data fetching` - ~81%
- `Navigation` - ~78%
- `Styling` - ~75%
- `UI components libraries` - ~71%
- `Graphics and animations` - ~70%
- `React Native features` - ~69%
- `Other app aspects` - ~69%
- `Deployment` - ~67%
- `Tools` - ~67%
- `Monorepo tools` - ~67%
- `React Native alternatives` - ~67%
- `Resources` - ~66%
- `Opinions` - ~66%

Based on that, we can adjust the order of sections, but the only viable alternative is to put `About you` as the first one, just like @kelset suggested 🤔 Personally I would rather follow the Devographics's path (they have a lot more experience and data on that topic) and keep the `About you` section as the last one. cc @kmagiera, @tsapeta, @kacperkapusciak

**Comment by bbarthec** [comment]
As of today, I'm closing this conversation as previously announced.
Thank you all for taking the time to help shape the survey! 💯 🙇 🎉 

The preview of the survey will be live until next Monday when we launch the production version 🚀 

Once again, thank you so much for your commitment to this initiative! ❤️
